File: src/main/java/com/enderzombi102/cmt/zone/ZoneManager.java
package com.enderzombi102.cmt.zone;

import net.fabricmc.fabric.api.util.NbtType;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtElement;
import net.minecraft.nbt.NbtList;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.world.World;
import org.apache.commons.lang3.RandomStringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class ZoneManager implements ZoneComponent {

	private final ServerWorld world;
	private final ArrayList< AbstractZone<? extends Entity> > zones = new ArrayList<>();

	public ZoneManager(World world) {
		if (world instanceof ServerWorld) this.world = (ServerWorld) world;
		else this.world = null;
	}

	@Override
	public void writeToNbt(@NotNull NbtCompound tag) {
		NbtList list = new NbtList();

		for (int i = 0; i < this.zones.size(); i++ ) {
			if ( this.zones.get(i).deleteOnExit ) continue;
			NbtCompound compoundTag = new NbtCompound();
			this.zones.get(i).writeToNbt(compoundTag);
			list.addElement(i, compoundTag);
		}

		tag.put("zones", list);
	}

	@Override
	@SuppressWarnings("ConstantConditions")
	public void readFromNbt(NbtCompound tag) {
		this.zones.clear();
		for ( NbtElement ctag : ( (NbtList) tag.get("zones") ).toArray( new NbtElement[0] ) ) {

			NbtCompound compoundTag = (NbtCompound) ctag;
			AbstractZone<? extends Entity> zone;

			switch ( compoundTag.getString("type") ) {
				case "FunctionZone":
					zone = new FunctionZone(this.world, compoundTag);
					break;
				default:  // this includes the "CommandZone" class/zone
					zone = new CommandZone(this.world, compoundTag);
			}

			this.zones.add( zone );
		}
	}

	@Override
	public void serverTick() {
		for ( AbstractZone<? extends Entity> zone : this.zones) {
			zone.tick();
		}
	}

	// ZoneComponent-specific methods

	@Override
	public void addZone(AbstractZone<? extends Entity> zone) {
		zone.world = this.world;
		if (this.world != null) this.zones.add(zone);
	}

	@Override
	public String removeZone(PlayerEntity entity) {
		for (AbstractZone<? extends Entity> zone : this.zones) {
			if( zone.getBox().intersects( entity.getBoundingBox() ) ) {
				String buf = zone.id;
				this.zones.remove( zone );
				return buf;
			}
		}
		return null;
	}

	@Override
	public void removeZone(String id) {
		this.zones.remove( this.getZone(id) );
	}

	@Override
	public boolean containsZone(String id) {
		return this.zones.stream().anyMatch(zone -> zone.id.equals(id) );
	}

	@Override
	public boolean entityInZone(List<Entity> entities, String id) {
		AbstractZone<? extends Entity> zone = this.getZone(id);
		if ( zone == null ) return false;
		for (Entity ent : entities ) {
			if ( zone.getLastEntities().contains(ent) ) return true;
		}
		return false;
	}

	@Override
	public @Nullable AbstractZone<? extends Entity> getZone(String id) {
		return this.zones.stream().filter( zone -> zone.id.equals(id) ).findAny().orElse(null);
	}

	@Override
	public List< AbstractZone<? extends Entity> > getZones() {
		return this.zones;
	}

	@Override
	public boolean isIdValid(String id) {
		return ! this.containsZone(id);
	}

	@Override
	public String getRandomId() {
		String id;
		do {
			id = RandomStringUtils.random(8);
		} while ( this.isIdValid(id) );
		return id;
	}

}

File: src/main/kotlin/com/enderzombi102/cmt/zone/ZoneManager.kt
package com.enderzombi102.cmt.zone

import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtList
import net.minecraft.server.world.ServerWorld
import net.minecraft.world.World
import org.apache.commons.lang3.RandomStringUtils

class ZoneManager( world: World? ) : ZoneComponent {
	private val world: ServerWorld? = if ( world is ServerWorld ) world else null
	private val zones = ArrayList<AbstractZone<out Entity>>()

	override fun writeToNbt(tag: NbtCompound) {
		val list = NbtList()
		for (i in zones.indices) {
			if (zones[i].deleteOnExit) continue
			val compoundTag = NbtCompound()
			zones[i].writeToNbt(compoundTag)
			list.addElement(i, compoundTag)
		}
		tag.put("zones", list)
	}

	override fun readFromNbt(tag: NbtCompound) {
		zones.clear()
		for (ctag in (tag["zones"] as NbtList?)!!.toTypedArray()) {
			val compoundTag = ctag as NbtCompound
			var zone: AbstractZone<out Entity>
			zone = when (compoundTag.getString("type")) {
				"FunctionZone" -> FunctionZone(world, compoundTag)
				else -> CommandZone(world, compoundTag)
			}
			zones.add(zone)
		}
	}

	override fun serverTick() {
		for (zone in zones) {
			zone.tick()
		}
	}

	// ZoneComponent-specific methods
	override fun addZone(zone: AbstractZone<out Entity>) {
		zone.world = world
		if (world != null) zones.add(zone)
	}

	override fun removeZone(entity: PlayerEntity): String? {
		for (zone in zones) {
			if (zone.box.intersects(entity.boundingBox)) {
				val buf = zone.id
				zones.remove(zone)
				return buf
			}
		}
		return null
	}

	override fun removeZone(id: String) {
		zones.remove(getZone(id))
	}

	override fun containsZone(id: String): Boolean {
		return zones.stream().anyMatch { zone: AbstractZone<out Entity>? -> zone!!.id == id }
	}

	override fun entityInZone(entities: List<Entity>, id: String): Boolean {
		val zone = getZone(id) ?: return false
		for (ent in entities) {
			if (zone.getLastEntities().contains(ent)) return true
		}
		return false
	}

	override fun getZone(id: String): AbstractZone<out Entity>? {
		return zones.stream().filter { zone: AbstractZone<out Entity>? -> zone!!.id == id }.findAny().orElse(null)
	}

	override fun getZones(): List<AbstractZone<out Entity>> {
		return zones
	}

	override fun isIdValid(id: String): Boolean {
		return !containsZone(id)
	}

	override val randomId: String
		get() {
			var id: String
			do {
				id = RandomStringUtils.random(8)
			} while (isIdValid(id))
			return id
		}
}
